packages/browser-sdk/src/hooksManager.ts

```diff
@@ -0,0 +1,105 @@
+import { CheckEvent, RawFeatures } from "./feature/features";
+import { CompanyContext, UserContext } from "./context";
+
+/**
+ * Hook for check event.
+ */
+export type CheckHook = {
+  type: "check-is-enabled" | "check-config";
+  callback: (checkEvent: CheckEvent) => void;
+};
+
+/**
+ * Hook for company update event.
+ */
+export type CompanyHook = {
+  type: "company";
+  callback: (company: CompanyContext) => void;
+};
+
+/**
+ * Hook for user update event.
+ */
+export type UserHook = {
+  type: "user";
+  callback: (user: UserContext) => void;
+};
+
+/**
+ * Hook for features updated event.
+ */
+export type FeaturesUpdatedHook = {
+  type: "features-updated";
+  callback: (features: RawFeatures) => void;
+};
+
+type trackEvent = {
+  user: UserContext;
+  company?: CompanyContext;
+  eventName: string;
+  attributes?: Record<string, any> | null;
+};
+
+/**
+ * Hook for track event.
+ */
+export type TrackHook = {
+  type: "track";
+  callback: (trackEvent: trackEvent) => void;
+};
+
+/**
+ * Hook definition.
+ */
+export type Hook =
+  | CheckHook
+  | FeaturesUpdatedHook
+  | UserHook
+  | CompanyHook
+  | TrackHook;
+
+/**
+ * Hooks manager.
+ * @internal
+ */
+export class HooksManager {
+  private hooks: {
+    "check-config": CheckHook[];
+    "check-is-enabled": CheckHook[];
+    "features-updated": FeaturesUpdatedHook[];
+    user: UserHook[];
+    company: CompanyHook[];
+    track: TrackHook[];
+  } = {
+    "check-config": [],
+    "check-is-enabled": [],
+    "features-updated": [],
+    user: [],
+    company: [],
+    track: [],
+  };
+
+  addHook(hook: Hook): void {
+    (this.hooks[hook.type] as Hook[]).push(hook);
+  }
+
+  triggerCheck(checkEvent: CheckEvent): void {
```

Comment:
I think you only need one function to trigger them: trigger(name: Hook["type"], payload: Hook[name]) ... something like this would add the type checking

Reply:
`trigger(name: Hook["type"], payload: Hook[name])` doesn't work, and I didn't find a way to make those types work unfortunately. If you have ideas I'm open for it!

Reply:
There's a couple of ways. One is to use Extract<Hook, { type }>. This should give you the union member that matches that given piece of the type (where type is your type parameter. I think your biggest issue is that type is not predefined so typescript sees it as 'string' instead of an union of those possible string costs

Reply:
Updated, let me know what you think!